tests/InferTypesTest.elm

```diff
@@ -0,0 +1,59 @@
+module InferTypesTest exposing (typeInference)
+
+import AST.Canonical as Canonical
+import AST.Common.Literal exposing (Literal(..))
+import AST.Common.Type as Type exposing (Type)
+import AST.Typed as Typed
+import Common
+import Dict.Any
+import Error exposing (TypeError)
+import Expect exposing (Expectation)
+import Stage.InferTypes
+import Test exposing (Test, describe, test)
+
+
+typeInference : Test
+typeInference =
+    let
+        runSection ( description, tests ) =
+            describe description
+                (List.map runTest tests)
+
+        runTest ( description, input, output ) =
+            test description <|
+                \() ->
+                    Stage.InferTypes.inferExpr input
+                        |> expectEqualInferType output
+    in
+    describe "Stage.InferType"
+        (List.map runSection
+            [ ( "list"
+              , [ ( "different types"
+                  , Canonical.List [ Canonical.Literal (Int 1), Canonical.Literal (String "two") ]
+                  , Err (Error.TypeMismatch Type.Int Type.String)
+                  )
+                , ( "more items types"
+                  , Canonical.List [ Canonical.Literal (Int 2), Canonical.Literal (Int 3), Canonical.Literal (Bool False) ]
+                  , Err (Error.TypeMismatch Type.Int Type.Bool)
+                  )
+                ]
+              )
+            ]
+        )
+
+
+expectEqualInferType :
+    Result TypeError Typed.Expr
+    -> Result TypeError Typed.Expr
+    -> Expectation
+expectEqualInferType expected actual =
```

Comment:
This function is a bit puzzling to me. Why does it automatically fail if `actual` is `Err`? What if `expected` is `Err` too?